File: julia/2020/day3.jl
using Test

function at(arr, idx)
    return arr[((idx-1)%length(arr))+1]
end

function part_one(input, direction = [3, 1])
    map = split.(split(input, "\n"), "")
    col = 1
    return reduce((v, row) -> begin
            v += at(row, col) == "#" ? 1 : 0
            col += direction[1]
            return v
        end, map[((1:end).-1).%direction[2].==0]; init = 0)
end

function part_two(input)
    return prod(map(f -> part_one(input, f), [[1, 1], [3, 1], [5, 1], [7, 1], [1, 2]]))
end

function test()
    example = """..##.......
#...#...#..
.#....#..#.
..#.#...#.#
.#...##..#.
..#.##.....
.#.#.#....#
.#........#
#.##...#...
#...##....#
.#..#...#.#"""

    @test part_one(example) == 7
    @test part_two(example) == 336
end

File: ruby/2023/day3.rb
require './commons/assert'

def neighbours(x, y, s, max_x, max_y)
  neighb = (y - 1..y + s).map { |n| [x - 1, n] }
  neighb << [x, y - 1]
  neighb << [x, y + s]
  neighb += (y - 1..y + s).map { |n| [x + 1, n] }
  neighb.filter { |n| n[0] >= 0 && n[0] < max_x && n[1] >= 0 && n[1] < max_y }
end

def process(lines)
  lines.each_with_index do |l, i|
    numbers = l.scan(/\d+/)
    numbers.each do |n|
      number = n.to_i
      idx = l.index n
      size = n.length

      # Remove already parsed numbers so repeated numbers in the same line can be found
      l.sub! n, '.' * size

      neighbours(i, idx, size, lines.size, l.size).each do |neigh|
        yield neigh, number
      end
    end
  end
end

def part_one(input)
  lines = input.split("\n")
  sum = 0

  process(lines) do |neigh, number|
    c = lines[neigh[0]][neigh[1]]
    next unless c != '.' && (c.ord < '0'.ord || c.ord > '9'.ord)

    sum += number
  end

  sum
end

def part_two(input)
  lines = input.split("\n")
  gears = {}

  process(lines) do |neigh, number|
    c = lines[neigh[0]][neigh[1]]
    if c == '*'
      if gears.key?(neigh)
        gears[neigh] << number
      else
        gears[neigh] = [number]
      end
    end
  end

  gears.map { |g| g[1].length == 2 ? g[1][0] * g[1][1] : 0 }.sum
end

def test
  example = %(467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592..20.
......755.
...$.*....
.664.598..)

  assert part_one(example), 4361
  assert part_two(example), 467_835
  0
end
